### PR TITLE
Debounce autocomplete

### LINKED ISSUE #34330: Better debounce the SQL editor
### Describe the bug

Same as https://github.com/metabase/metabase/issues/34001, but with SQL queries. First letter will fire a query, and this is a problem in https://github.com/metabase/metabase/issues/30588

### To Reproduce

Just create a new sql question and type a letter, see the endpoint being hit

### Expected behavior

We should expect the user finishes typing before hitting the endpoint

### Logs

_No response_

### Information about your Metabase installation

```JSON
47.3
```

### Severity

P2

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by romeovs**
Closes https://github.com/metabase/metabase/issues/34330

### Description

The Ace Editor triggers an autocompletion when the user starts typing initially, which isn't caught the debounced `_retriggerAutocomplete` function.

This PR adds a case to avoid calling the API when this happens. The autocomplete will still happen when the user stops typing for a while, just like before.

### How to verify

Describe the steps to verify that the changes are working as expected.

1. `New` → `SQL Query` → `Sample Database`
2. Open the browser network tab
3. Start typing: `SELE` in one go and verify only a request to `/api/database/1/autocomplete_suggestions?substring=SELE` is made, not `/api/database/1/autocomplete_suggestions?substring=S`.
4. Clear the editor and type `S`, verify that a request to `/api/database/1/autocomplete_suggestions?substring=S` is made, but only after about 700ms after you stop typing.

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Review by kamilmielnik — COMMENTED** [review]
Tests look good but the fix looks a bit complex.

I don't fully understand it yet, I will debug it on Monday, but I'm leaving some early comments.

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx [hunk 1]
```diff
@@ -455,15 +454,30 @@ export class NativeQueryEditor extends Component<
         }
 
         try {
-          let { results, timestamp } = this._lastAutoComplete;
+          let skip = false;
+          if (prefix.length <= 1 && prefix !== lastAutoComplete.prefix) {
+            // ACE triggers an autocomplete when the user starts typing that is
+            // not caught by debouncing _retriggerAutocomplete.
```
[the remark this answers is not in the document]
Comment (kamilmielnik) [reply frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx]: > I am happy to give debouncing the autocompleter another go but don't think we should spend much more time on this.

Let's do that next time we have to touch this part :+1:

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx [hunk 2]
```diff
@@ -455,15 +454,30 @@ export class NativeQueryEditor extends Component<
...
+            // Here we prevent it from running initially (iff the prefix is just on character).
+            // It will run eventually, even if the prefix is only on character, after the user stops typing.
```
[the remark this answers is not in the document]
Comment (romeovs) [reply frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx]: The autocompleter gets triggered from `_retriggerAutocomplete`, which triggers when the user stops typing (and which is the expected behavior).

The problem was that ACE itself was triggering the autocompleter too, which this fix prevents by ignoring that autocompletion request.

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx [hunk 3]
```diff
@@ -455,15 +460,32 @@ export class NativeQueryEditor extends Component<
...
+            // Here we prevent it from actually calling the autocomplete endpoint.
+            // It will run eventually even if the prefix is only one character,
+            // after the user stops typing, because _retriggerAutocomplete will get called with the same prefix.
+            lastAutoComplete = {
+              timestamp: 0,
+              prefix,
+              results: lastAutoComplete.results,
+            };
+
+            callback(null, []);
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx:475]: I think we want to call the callback with `lastAutoComplete.results` mapped like in line 530, not an empty array.

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx [hunk 4]
```diff
@@ -455,15 +460,32 @@ export class NativeQueryEditor extends Component<
         }
 
         try {
-          let { results, timestamp } = this._lastAutoComplete;
+          if (prefix.length <= 1 && prefix !== lastAutoComplete.prefix) {
```
[the remark this answers is not in the document]
Comment (romeovs) [reply frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx]: I've added a fix for this and also added a test case for this regression.

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx [hunk 5]
```diff
@@ -440,7 +442,19 @@ export class NativeQueryEditor extends Component<
...
+      results: [],
+      first: true,
+    };
+
+    const resultsForAce = (results: [string, string][]) =>
+      results.map(([name, meta]) => ({
+        name: name,
+        value: name,
+        meta: meta,
+      })) ?? [];
```
Comment (uladzimirdev) [inline frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx:457]: I doubt `?? []` will ever be called
```suggestion
      }));
```

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx [hunk 6]
```diff
@@ -455,18 +469,37 @@ export class NativeQueryEditor extends Component<
         }
 
         try {
-          let { results, timestamp } = this._lastAutoComplete;
+          if (lastAutoComplete.first && prefix !== lastAutoComplete.prefix) {
```
Comment (uladzimirdev) [inline frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx:472]: to my mind `first` value is the same as `timestamp === 0`, am I wrong?

File: frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx [hunk 7]
```diff
@@ -505,12 +538,7 @@ export class NativeQueryEditor extends Component<
           }
 
           // transform results into what ACE expects
-          const resultsForAce = results?.map(([name, meta]) => ({
-            name: name,
-            value: name,
-            meta: meta,
-          }));
-          callback(null, resultsForAce || []);
+          callback(null, resultsForAce(results));
```
Comment (uladzimirdev) [inline frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx:541]: I'd rename `resultsForAce` to highlight it's a method, maybe `getResultsForAce` or `prepareResultsForAce`

**Review by uladzimirdev — COMMENTED** [review]
Please check comments inline

**Review by uladzimirdev — APPROVED** [review]
LGTM

**Pull request merged**